File: backend/mcp_server.py
#!/usr/bin/env python3
import sys
import os
from contextlib import contextmanager
from datetime import date
from typing import Optional

sys.path.insert(0, os.path.dirname(__file__))  # makes `app` importable

from mcp.server.fastmcp import FastMCP
from app.database import SessionLocal
from app.models import (
    Profile, Company, Role, Skill, SkillType,
    Education, Certification, JobApplication, ApplicationStatus,
    Accomplishment, RoleSkill, Award, Presentation, TeamStructure, OrgPosition,
    Responsibility, AccomplishmentCategory,
)
from app.services.export_service import generate_export_document

mcp = FastMCP("career-manager")


@contextmanager
def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


def _date_str(d) -> Optional[str]:
    return d.isoformat() if d else None


def _profile_dict(p) -> dict:
    return {
        "id": p.id,
        "name": p.name,
        "email": p.email,
        "phone": p.phone,
        "city": p.city,
        "state": p.state,
        "portfolio_url": p.portfolio_url,
        "github_url": p.github_url,
        "linkedin_url": p.linkedin_url,
    }


def _company_dict(c) -> dict:
    return {
        "id": c.id,
        "name": c.name,
        "description": c.description,
        "industry": c.industry,
        "start_date": _date_str(c.start_date),
        "end_date": _date_str(c.end_date),
        "city": c.city,
        "state": c.state,
    }


def _role_summary_dict(r, company_name: str = None) -> dict:
    return {
        "id": r.id,
        "company_id": r.company_id,
        "company_name": company_name or (r.company.name if r.company else None),
        "title": r.title,
        "level": r.level,
        "start_date": _date_str(r.start_date),
        "end_date": _date_str(r.end_date),
        "city": r.city,
        "state": r.state,
        "is_remote": r.is_remote,
        "summary": r.summary,
    }


def _role_detail_dict(r) -> dict:
    d = _role_summary_dict(r)
    d["accomplishments"] = [
        {
            "id": a.id,
            "description": a.description,
            "tech_stack": a.tech_stack,
            "key_result": a.key_result,
            "category": a.category.value if a.category else None,
            "sort_order": a.sort_order,
        }
        for a in sorted(r.accomplishments, key=lambda x: x.sort_order or 0)
    ]
    d["responsibilities"] = [
        {
            "id": resp.id,
            "description": resp.description,
            "sort_order": resp.sort_order,
        }
        for resp in sorted(r.responsibilities, key=lambda x: x.sort_order or 0)
    ]
    d["skills"] = [
        {
            "skill_id": rs.skill_id,
            "name": rs.skill.name if rs.skill else None,
            "skill_type": rs.skill.skill_type.value if rs.skill and rs.skill.skill_type else None,
            "proficiency": rs.proficiency,
        }
        for rs in r.role_skills
    ]
    d["awards"] = [
        {
            "id": a.id,
            "title": a.title,
            "description": a.description,
            "date": _date_str(a.date),
            "issuer": a.issuer,
            "resume_relevant": a.resume_relevant,
        }
        for a in r.awards
    ]
    d["presentations"] = [
        {
            "id": p.id,
            "title": p.title,
            "venue": p.venue,
            "date": _date_str(p.date),
            "audience": p.audience,
            "resume_relevant": p.resume_relevant,
        }
        for p in r.presentations
    ]
    d["team_structure"] = (
        {
            "direct_reports": r.team_structure.direct_reports,
            "team_size": r.team_structure.team_size,
            "responsibilities": r.team_structure.responsibilities,
        }
        if r.team_structure
        else None
    )
    d["org_position"] = (
        {
            "reports_to": r.org_position.reports_to,
            "department": r.org_position.department,
            "org_level": r.org_position.org_level,
        }
        if r.org_position
        else None
    )
    return d


def _application_dict(app) -> dict:
    return {
        "id": app.id,
        "company_name": app.company_name,
        "job_title": app.job_title,
        "req_url": app.req_url,
        "date_applied": _date_str(app.date_applied),
        "status": app.status.value if app.status else None,
        "status_date": _date_str(app.status_date),
        "cover_letter_submitted": app.cover_letter_submitted,
        "notes": app.notes,
        "created_at": app.created_at.isoformat() if app.created_at else None,
        "updated_at": app.updated_at.isoformat() if app.updated_at else None,
    }


# ─── Read Tools ────────────────────────────────────────────────────────────────

@mcp.tool()
def get_profile() -> dict:
    """Get the user's contact info, LinkedIn, GitHub, and portfolio URL."""
    with get_db() as db:
        profile = db.query(Profile).first()
        if not profile:
            return {"error": "No profile found"}
        return _profile_dict(profile)


@mcp.tool()
def list_companies() -> list[dict]:
    """List all employers with their IDs, dates, and industry."""
    with get_db() as db:
        companies = db.query(Company).order_by(Company.start_date.desc()).all()
        return [_company_dict(c) for c in companies]


@mcp.tool()
def list_all_roles() -> list[dict]:
    """List every role across all companies, newest first. Includes company_name."""
    with get_db() as db:
        roles = (
            db.query(Role)
            .join(Company, Role.company_id == Company.id)
            .order_by(Role.start_date.desc())
            .all()
        )
        return [_role_summary_dict(r) for r in roles]


@mcp.tool()
def list_roles_for_company(company_id: int) -> list[dict]:
    """List all roles at a specific company by company_id."""
    with get_db() as db:
        roles = (
            db.query(Role)
            .filter(Role.company_id == company_id)
            .order_by(Role.start_date.desc())
            .all()
        )
        return [_role_summary_dict(r) for r in roles]


@mcp.tool()
def get_role_detail(role_id: int) -> dict:
    """Get full detail for a role: accomplishments, responsibilities, skills, awards, presentations, team structure."""
    with get_db() as db:
        role = db.query(Role).filter(Role.id == role_id).first()
        if not role:
            return {"error": f"Role {role_id} not found"}
        return _role_detail_dict(role)


@mcp.tool()
def export_career_context(
    role_ids: list[int],
    include_supporting: bool = True,
    include_awards: bool = True,
    include_presentations: bool = True,
    include_responsibilities: bool = True,
) -> str:
    """Export full career context as a markdown document for the given role IDs.

    Pass role_ids from list_all_roles(). Set include_supporting=True to include
    additional context bullets beyond resume bullets.
    """
    with get_db() as db:
        return generate_export_document(
            db=db,
            role_ids=role_ids,
            include_supporting=include_supporting,
            include_awards=include_awards,
            include_presentations=include_presentations,
            include_responsibilities=include_responsibilities,
        )


@mcp.tool()
def list_skills(skill_type: Optional[str] = None) -> list[dict]:
    """List all skills. Optionally filter by skill_type: 'technical' or 'soft'."""
    with get_db() as db:
        query = db.query(Skill)
        if skill_type:
            try:
                st = SkillType(skill_type.lower())
                query = query.filter(Skill.skill_type == st)
            except ValueError:
                return [{"error": f"Invalid skill_type '{skill_type}'. Use 'technical' or 'soft'."}]
        skills = query.order_by(Skill.name).all()
        return [
            {"id": s.id, "name": s.name, "skill_type": s.skill_type.value if s.skill_type else None}
            for s in skills
        ]


@mcp.tool()
def list_education() -> list[dict]:
    """List all education records."""
    with get_db() as db:
        education = db.query(Education).order_by(Education.start_date.desc()).all()
        return [
            {
                "id": e.id,
                "institution": e.institution,
                "degree": e.degree,
                "field": e.field,
                "city": e.city,
                "state": e.state,
                "start_date": _date_str(e.start_date),
                "end_date": _date_str(e.end_date),
            }
            for e in education
        ]


@mcp.tool()
def list_certifications() -> list[dict]:
    """List all certification records."""
    with get_db() as db:
        certs = db.query(Certification).order_by(Certification.issue_date.desc()).all()
        return [
            {
                "id": c.id,
                "name": c.name,
                "issuing_org": c.issuing_org,
                "issue_date": _date_str(c.issue_date),
                "expiry_date": _date_str(c.expiry_date),
                "credential_id": c.credential_id,
            }
            for c in certs
        ]


# ─── Write Tools ───────────────────────────────────────────────────────────────

@mcp.tool()
def create_job_application(
    company_name: str,
    job_title: str,
    date_applied: str,
    req_url: Optional[str] = None,
    cover_letter_submitted: bool = False,
    notes: Optional[str] = None,
) -> dict:
    """Log a new job application.

    date_applied must be in ISO format: YYYY-MM-DD.
    """
    with get_db() as db:
        app = JobApplication(
            company_name=company_name,
            job_title=job_title,
            date_applied=date.fromisoformat(date_applied),
            req_url=req_url,
            status=ApplicationStatus.APPLIED,
            cover_letter_submitted=cover_letter_submitted,
            notes=notes,
        )
        db.add(app)
        db.commit()
        db.refresh(app)
        return _application_dict(app)


@mcp.tool()
def list_job_applications(status_filter: Optional[str] = None) -> list[dict]:
    """List job applications, optionally filtered by status.

    Valid statuses: applied, email_rejected, ghosted, withdrew, req_pulled,
    interview_scheduled, interview_rejected, offered, accepted.
    """
    with get_db() as db:
        query = db.query(JobApplication)
        if status_filter:
            try:
                st = ApplicationStatus(status_filter.lower())
                query = query.filter(JobApplication.status == st)
            except ValueError:
                return [{"error": f"Invalid status '{status_filter}'."}]
        apps = query.order_by(JobApplication.date_applied.desc()).all()
        return [_application_dict(a) for a in apps]


@mcp.tool()
def get_job_application(application_id: int) -> dict:
    """Get a single job application by ID."""
    with get_db() as db:
        app = db.query(JobApplication).filter(JobApplication.id == application_id).first()
        if not app:
            return {"error": f"Application {application_id} not found"}
        return _application_dict(app)


@mcp.tool()
def update_application_status(
    application_id: int,
    status: str,
    status_date: Optional[str] = None,
) -> dict:
    """Update the status of a job application.

    Valid statuses: applied, email_rejected, ghosted, withdrew, req_pulled,
    interview_scheduled, interview_rejected, offered, accepted.
    status_date must be ISO format YYYY-MM-DD if provided, otherwise today's date is used.
    """
    with get_db() as db:
        app = db.query(JobApplication).filter(JobApplication.id == application_id).first()
        if not app:
            return {"error": f"Application {application_id} not found"}
        try:
            new_status = ApplicationStatus(status.lower())
        except ValueError:
            return {"error": f"Invalid status '{status}'."}
        app.status = new_status
        app.status_date = date.fromisoformat(status_date) if status_date else date.today()
        db.commit()
        db.refresh(app)
        return _application_dict(app)


@mcp.tool()
def update_application_notes(application_id: int, notes: str) -> dict:
    """Update the notes field on a job application (recruiter info, salary, feedback, etc.)."""
    with get_db() as db:
        app = db.query(JobApplication).filter(JobApplication.id == application_id).first()
        if not app:
            return {"error": f"Application {application_id} not found"}
        app.notes = notes
        db.commit()
        db.refresh(app)
        return _application_dict(app)


if __name__ == "__main__":
    mcp.run()  # stdio transport

File: server/mcp_server.ts
#!/usr/bin/env tsx
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { z } from 'zod';
import { eq, desc, asc, inArray } from 'drizzle-orm';
import { db } from './src/db/index.js';
import {
  profiles, companies, roles, skills, education, certifications,
  job_applications, role_skills, accomplishments, responsibilities,
  awards, presentations, team_structures, org_positions, knowledge_entries,
} from './src/db/schema.js';
import { generateExportDocument } from './src/services/exportService.js';
import { getInstructions } from './src/services/instructionsService.js';
import { convertYamlToWordBytes } from './src/services/yamlToWord.js';
import { persistResumeOutput } from './src/services/outputPersistenceService.js';

const server = new McpServer({ name: 'career-manager', version: '1.0.0' });

// ── Read Tools ────────────────────────────────────────────────────────────────

server.tool('get_profile', 'Get the user\'s contact info, LinkedIn, GitHub, and portfolio URL.', {}, async () => {
  const profile = db.select().from(profiles).limit(1).all()[0] ?? null;
  return { content: [{ type: 'text' as const, text: JSON.stringify(profile) }] };
});

server.tool('list_companies', 'List all employers with their IDs, dates, and industry.', {}, async () => {
  const rows = db.select().from(companies).orderBy(desc(companies.start_date)).all();
  return { content: [{ type: 'text' as const, text: JSON.stringify(rows) }] };
});

server.tool('list_all_roles', 'List every role across all companies, newest first. Includes company_name.', {}, async () => {
  const roleRows = db.select().from(roles).orderBy(desc(roles.start_date)).all();
  const result = roleRows.map(r => {
    const company = db.select().from(companies).where(eq(companies.id, r.company_id)).limit(1).all()[0];
    return { ...r, company_name: company?.name ?? null };
  });
  return { content: [{ type: 'text' as const, text: JSON.stringify(result) }] };
});

server.tool(
  'list_roles_for_company',
  'List all roles at a specific company.',
  { company_id: z.number().describe('The company ID') },
  async ({ company_id }) => {
    const rows = db.select().from(roles)
      .where(eq(roles.company_id, company_id))
      .orderBy(desc(roles.start_date))
      .all();
    return { content: [{ type: 'text' as const, text: JSON.stringify(rows) }] };
  },
);

server.tool(
  'get_role_detail',
  'Get full detail for a role: accomplishments, responsibilities, skills, awards, presentations, team structure.',
  { role_id: z.number().describe('The role ID') },
  async ({ role_id }) => {
    const role = db.select().from(roles).where(eq(roles.id, role_id)).limit(1).all()[0];
    if (!role) return { content: [{ type: 'text' as const, text: JSON.stringify({ error: `Role ${role_id} not found` }) }] };

    const company = db.select().from(companies).where(eq(companies.id, role.company_id)).limit(1).all()[0];
    const accs = db.select().from(accomplishments).where(eq(accomplishments.role_id, role_id)).orderBy(asc(accomplishments.sort_order)).all();
    const resps = db.select().from(responsibilities).where(eq(responsibilities.role_id, role_id)).orderBy(asc(responsibilities.sort_order)).all();
    const rs = db.select().from(role_skills).where(eq(role_skills.role_id, role_id)).all().map(rs => {
      const skill = db.select().from(skills).where(eq(skills.id, rs.skill_id)).limit(1).all()[0];
      return { ...rs, skill_name: skill?.name, skill_type: skill?.skill_type };
    });
    const awardRows = db.select().from(awards).where(eq(awards.role_id, role_id)).all();
    const presRows = db.select().from(presentations).where(eq(presentations.role_id, role_id)).all();
    const ts = db.select().from(team_structures).where(eq(team_structures.role_id, role_id)).limit(1).all()[0] ?? null;
    const op = db.select().from(org_positions).where(eq(org_positions.role_id, role_id)).limit(1).all()[0] ?? null;

    const detail = {
      ...role,
      company_name: company?.name ?? null,
      accomplishments: accs,
      responsibilities: resps,
      skills: rs,
      awards: awardRows,
      presentations: presRows,
      team_structure: ts,
      org_position: op,
    };

    return { content: [{ type: 'text' as const, text: JSON.stringify(detail) }] };
  },
);

server.tool(
  'export_career_context',
  'Export full career context as a markdown document for the given role IDs.',
  {
    role_ids: z.array(z.number()).describe('Role IDs to include'),
    include_supporting: z.boolean().optional().default(true),
    include_awards: z.boolean().optional().default(true),
    include_presentations: z.boolean().optional().default(true),
    include_responsibilities: z.boolean().optional().default(true),
  },
  async ({ role_ids, include_supporting, include_awards, include_presentations, include_responsibilities }) => {
    const markdown = generateExportDocument(
      role_ids,
      include_supporting ?? true,
      include_awards ?? true,
      include_presentations ?? true,
      include_responsibilities ?? true,
    );
    return { content: [{ type: 'text' as const, text: markdown }] };
  },
);

server.tool(
  'list_skills',
  "List all skills. Optionally filter by skill_type: 'technical' or 'soft'.",
  { skill_type: z.enum(['technical', 'soft']).optional().describe("Filter by 'technical' or 'soft'") },
  async ({ skill_type }) => {
    let rows = db.select().from(skills).orderBy(asc(skills.name)).all();
    if (skill_type) rows = rows.filter(s => s.skill_type?.toUpperCase() === skill_type.toUpperCase());
    return { content: [{ type: 'text' as const, text: JSON.stringify(rows) }] };
  },
);

server.tool('list_education', 'List all education records.', {}, async () => {
  const rows = db.select().from(education).orderBy(desc(education.start_date)).all();
  return { content: [{ type: 'text' as const, text: JSON.stringify(rows) }] };
});

server.tool('list_certifications', 'List all certification records.', {}, async () => {
  const rows = db.select().from(certifications).orderBy(desc(certifications.issue_date)).all();
  return { content: [{ type: 'text' as const, text: JSON.stringify(rows) }] };
});

server.tool(
  'get_instructions',
  'Get the AI evaluation instructions that define how to assess a job requisition against career history.',
  {},
  async () => {
    const content = getInstructions();
    const wrapped = `BINDING INSTRUCTIONS — you must follow every rule below exactly and in order. These are not suggestions; they define the required process, output format, and YAML schema for this session.\n\n${content}\n\nEND OF BINDING INSTRUCTIONS. Do not deviate from the process, output templates, or YAML schema defined above.`;
    return { content: [{ type: 'text' as const, text: wrapped }] };
  },
);

server.tool(
  'get_recent_roles_context',
  'Get the most recent N roles with full career context (accomplishments, responsibilities, awards, presentations, supporting details). Use this at the start of a job evaluation workflow.',
  { limit: z.number().int().min(1).max(20).default(5).describe('Number of most recent roles to include (default 5)') },
  async ({ limit }) => {
    const recentRoles = db.select().from(roles).orderBy(desc(roles.start_date)).limit(limit).all();
    const roleIds = recentRoles.map(r => r.id);
    if (roleIds.length === 0) return { content: [{ type: 'text' as const, text: 'No roles found.' }] };
    const markdown = generateExportDocument(roleIds, true, true, true, true);
    return { content: [{ type: 'text' as const, text: markdown }] };
  },
);

server.tool(
  'generate_resume',
  'Convert a YAML resume definition to a formatted Word document (.docx) and save it to disk. Returns the saved file path.',
  {
    yaml_content: z.string().describe('Full YAML resume content with resumeStructure key'),
    company_name: z.string().describe('Company name used in the output filename'),
  },
  async ({ yaml_content, company_name }) => {
    const [docBytes, filename] = await convertYamlToWordBytes(yaml_content, company_name);
    const [, relativePath] = persistResumeOutput(docBytes, filename);
    return { content: [{ type: 'text' as const, text: JSON.stringify({ success: true, path: relativePath, filename }) }] };
  },
);

// ── Write Tools ───────────────────────────────────────────────────────────────

server.tool(
  'create_job_application',
  'Log a new job application. date_applied must be YYYY-MM-DD.',
  {
    company_name: z.string(),
    job_title: z.string(),
    date_applied: z.string().describe('ISO date YYYY-MM-DD'),
    req_url: z.string().optional(),
    cover_letter_submitted: z.boolean().optional().default(false),
    notes: z.string().optional(),
  },
  async ({ company_name, job_title, date_applied, req_url, cover_letter_submitted, notes }) => {
    const now = new Date().toISOString().replace('T', ' ').split('.')[0] + '.000000';
    const [created] = await db.insert(job_applications).values({
      company_name,
      job_title,
      date_applied,
      req_url: req_url ?? null,
      status: 'applied',
      cover_letter_submitted: cover_letter_submitted ?? false,
      notes: notes ?? null,
      created_at: now,
      updated_at: now,
    }).returning();
    return { content: [{ type: 'text' as const, text: JSON.stringify(created) }] };
  },
);

server.tool(
  'list_job_applications',
  'List job applications, optionally filtered by status.',
  { status_filter: z.string().optional().describe('Filter by status value') },
  async ({ status_filter }) => {
    let rows = db.select().from(job_applications).orderBy(desc(job_applications.date_applied)).all();
    if (status_filter) rows = rows.filter(a => a.status === status_filter);
    return { content: [{ type: 'text' as const, text: JSON.stringify(rows) }] };
  },
);

server.tool(
  'get_job_application',
  'Get a single job application by ID.',
  { application_id: z.number() },
  async ({ application_id }) => {
    const app = db.select().from(job_applications).where(eq(job_applications.id, application_id)).limit(1).all()[0];
    if (!app) return { content: [{ type: 'text' as const, text: JSON.stringify({ error: `Application ${application_id} not found` }) }] };
    return { content: [{ type: 'text' as const, text: JSON.stringify(app) }] };
  },
);

server.tool(
  'update_application_status',
  'Update application status. Valid: applied, email_rejected, ghosted, withdrew, req_pulled, interview_scheduled, interview_rejected, offered, accepted.',
  {
    application_id: z.number(),
    status: z.string(),
    status_date: z.string().optional().describe('ISO date YYYY-MM-DD, defaults to today'),
  },
  async ({ application_id, status, status_date }) => {
    const now = new Date().toISOString().replace('T', ' ').split('.')[0] + '.000000';
    const [updated] = await db.update(job_applications)
      .set({
        status,
        status_date: status_date ?? new Date().toISOString().split('T')[0],
        updated_at: now,
      })
      .where(eq(job_applications.id, application_id))
      .returning();
    if (!updated) return { content: [{ type: 'text' as const, text: JSON.stringify({ error: `Application ${application_id} not found` }) }] };
    return { content: [{ type: 'text' as const, text: JSON.stringify(updated) }] };
  },
);

server.tool(
  'update_application_notes',
  'Update the notes field on a job application.',
  {
    application_id: z.number(),
    notes: z.string().describe('Recruiter info, salary, feedback, etc.'),
  },
  async ({ application_id, notes }) => {
    const now = new Date().toISOString().replace('T', ' ').split('.')[0] + '.000000';
    const [updated] = await db.update(job_applications)
      .set({ notes, updated_at: now })
      .where(eq(job_applications.id, application_id))
      .returning();
    if (!updated) return { content: [{ type: 'text' as const, text: JSON.stringify({ error: `Application ${application_id} not found` }) }] };
    return { content: [{ type: 'text' as const, text: JSON.stringify(updated) }] };
  },
);

// ── Start ─────────────────────────────────────────────────────────────────────

const transport = new StdioServerTransport();
await server.connect(transport);
